File: src/main/java/com/billerby/roadvault/service/EmailService.java
package com.billerby.roadvault.service;

import com.billerby.roadvault.dto.InvoiceDTO;
import freemarker.template.TemplateException;
import jakarta.mail.MessagingException;

import java.io.IOException;
import java.util.List;

/**
 * Service interface for sending emails.
 */
public interface EmailService {
    
    /**
     * Send an invoice email to a recipient.
     *
     * @param invoiceDTO The invoice data to use
     * @param pdfAttachment PDF attachment to include with the email
     * @throws MessagingException If there is an error sending the message
     * @throws IOException If there is an error processing templates
     * @throws TemplateException If there is an error in the template
     */
    void sendInvoiceEmail(InvoiceDTO invoiceDTO, byte[] pdfAttachment) 
            throws MessagingException, IOException, TemplateException;
    
    /**
     * Send invoices to all recipients in the list.
     *
     * @param invoiceIds List of invoice IDs to send
     * @return Number of successfully sent emails
     */
    int sendInvoices(List<Long> invoiceIds);
    
    /**
     * Send invoices to all recipients for a specific billing.
     *
     * @param billingId The billing ID to send invoices for
     * @return Number of successfully sent emails
     */
    int sendInvoicesForBilling(Long billingId);
}

File: src/frontend/src/services/email.service.js
import api from './api';

/**
 * Service for managing email operations in the frontend.
 */
class EmailService {
  /**
   * Send a single invoice via email
   * @param {number} invoiceId Invoice ID
   * @returns {Promise} Promise with email send result
   */
  sendInvoiceEmail(invoiceId) {
    return api.post(`/v1/emails/invoices/${invoiceId}`);
  }

  /**
   * Send a reminder for a single invoice via email
   * @param {number} invoiceId Invoice ID
   * @returns {Promise} Promise with email send result
   */
  sendInvoiceReminder(invoiceId) {
    return api.post(`/v1/emails/invoices/${invoiceId}/reminder`);
  }

  /**
   * Send multiple invoices via email
   * @param {Array<number>} invoiceIds Array of invoice IDs
   * @returns {Promise} Promise with email send results
   */
  sendInvoiceEmails(invoiceIds) {
    return api.post('/v1/emails/invoices', { invoiceIds });
  }

  /**
   * Send all invoices for a billing via email
   * @param {number} billingId Billing ID
   * @returns {Promise} Promise with email send results
   */
  sendInvoicesForBilling(billingId) {
    return api.post(`/v1/emails/invoices/billing/${billingId}`);
  }
}

export default new EmailService();
